### ISSUE TITLE
Rounding issue in setupInternalCanvasMatrix()

### ISSUE THREAD

**Issue by rabross** [Dimezis/BlurView#128]
Version 1.6.5

There's a problem in `setupInternalCanvasMatrix()` where a rounding issue can occur causing `internalCanvas` and `rootView` sizes to not match. This leads to some blurring issues at the bottom of `BlurView`.

This is down to using `roundingScaleFactor` during `internalCanvas.scale()` for height instead of the 'real' height scale factor.

I've fixed on my side by producing:
`float nonRoundingScaleFactorHeight = (float) rootView.getHeight() / internalCanvas.getHeight();`
and using that for scaling:
`internalCanvas.scale(1 / scaleFactor, 1 / nonRoundingScaleFactorHeight);`

It might be worth looking into.

**Comment by Dimezis** [comment]
This will result in a slightly vertically stretched/squeezed image depending on the nonRoundingScaleFactorHeight.
Can you post a screenshot of what you're seeing? Also the size of the BlurView in that setup.

**Comment by rabross** [comment]
Thanks for getting back to me.
The (slightly hacky) solution above seems to be working well across devices.

Some screenshots without the fix. I've tried to highlight the problem

Root view capture with extra row of transparent pixels at bottom (i.e. root has no background and no frame clear drawable)
<img width="434" alt="pre-blur-capture" src="https://user-images.githubusercontent.com/3681815/98420612-791fe180-207f-11eb-996e-e1abc3e78102.png">

The problem is apparent in the post blur. The transparency has blurred into the image
<img width="431" alt="post-blur" src="https://user-images.githubusercontent.com/3681815/98420621-7f15c280-207f-11eb-9d3a-299fc320b74e.png">

<img width="485" alt="blurview-height" src="https://user-images.githubusercontent.com/3681815/98420626-81781c80-207f-11eb-8356-805839fd92c0.png">

<img width="487" alt="rootview-height" src="https://user-images.githubusercontent.com/3681815/98420631-84730d00-207f-11eb-9f59-7a1a4127aa06.png">

<img width="486" alt="canvas-height" src="https://user-images.githubusercontent.com/3681815/98420636-88069400-207f-11eb-9063-c68151ade395.png">

Here you can see the issue. This value is being round down to 204 when drawn to canvas which doesn't fill the 205 height of the canvas.
<img width="486" alt="rounding-issue" src="https://user-images.githubusercontent.com/3681815/98420638-8b018480-207f-11eb-90b7-cba3d94641e3.png">

**Comment by rabross** [comment]
Also in the above case:

`rootView.getWidth()` is 1080
`internalCanvas.getWidth()` is 192
`internalBitmap.getWidth()` is 192

**Comment by Dimezis** [comment]
I'm not sure what to do here TBH. 
It's obviously a bug and looks bad, however your fix is potentially squeezing/stretching the resulting image on the vertical axis.
On the blurred version it's not that noticeable, so I might go for it

**Comment by Dimezis** [comment]
Alternative approach would be to scale by `nonRoundingScaleFactorHeight` in both directions, so it's preserving the aspect ratio, but goes slightly behind the borders on X axis

**Comment by Mickey34** [comment]
Hi,
I've faced the same issue recently but suggested changes to setupInternalCanvasMatrix() haven't solved my problem. I had to add the same calculation into BlockingBlurController.draw() method to make it work:
```     
float nonRoundingScaleFactorHeight = (float) blurView.getHeight() / internalCanvas.getHeight();
canvas.save();
canvas.scale(scaleFactor, nonRoundingScaleFactorHeight);
canvas.drawBitmap(internalBitmap, 0, 0, paint);
canvas.restore();
```
I also wasn't able to use the solution with calculating the real rounding factor from **rootView**.getHeight(). It caused problems when rootView height is much more than blurView height. For example, nonRoundingScaleFactorHeight was 32 while scaleFactor was near 4.5 (don't remember exact number) for the case where rootView height is 1152 but blurView height is 96 pixels. As a result, my item wasn't blurred at all.
When I tried blurView instead of rootView for calculating nonRoundingScaleFactorHeight all issues disappeared.
